src/ocean/core/TypeConvert.d

```diff
@@ -245,3 +245,65 @@ unittest
     static assert (!is(typeof(arrayOf!(int*)(1000))));
     static assert (!is(typeof(arrayOf!(int)((int[]).init))));
 }
+
+
+/*******************************************************************************
+
+    Generates delegate that stores specified `context` as a delegate context
+    pointer and, when called, forwards it to function `F` as a regular argument.
+
+    Intended to be used as a performance optimization hack to create
+    no-allocation closures that only need to capture one pointer size argument.
+
+    Params:
+        F = function to call when delegate is called, must take exactly one
+            void* argument which is the passed context
+        context = context pointer to forward to F when resulting delegate is
+            called
+
+    Returns:
+        forged delegate that can be passed to any API expecting regular `void
+        delegate()`
+
+*******************************************************************************/
+
+void delegate() toContextDg ( alias F ) ( void* context )
+{
+    static assert (is(typeof(&F) == void function (void*)));
+
+    // Makes use of the fact that D ABI allows converting aggregate methods
+    // to delegates such that resulting delegate context pointer becomes
+    // aggregate `this`:
```

Comment:
It makes uses of a second fact, as well, which you use a bit later.

Suggested wording:
``
This code makes use of two facts:
(1) The D ABI allows aggregate methods to be converted to delegates, such that the delegate context pointer
becomes the `this` pointer of the aggregate; and 
(2) The compiler supports explicit modification of the .ptr member of a delegate, without modifying the existing .functptr.
``

The second fact is actually pretty interesting, because it means the compiler knows exactly what you're doing here. The only reason for supporting that syntax in the compiler, is to allow hacks like this. Admittedly the spec doesn't mention that you can do this, but it's explicitly coded in the compiler.

Reply:
Created fixup PR https://github.com/sociomantic-tsunami/ocean/pull/350